### PR TITLE
Converted Team 'grt123' identification and classification algorithms

### PR DESCRIPTION
**Pull request by dchansen**
## Description

This pull request implements the grt123 identification algorithm.
It is based on PyTorch, which currently conflicts with Tensorflow, so the pull request also disables the Keras import. 

While the algorithm is quite fast on GPU, it is painfully slow (but working) on the CPU due to a bug in PyTorch. The bug is fixed in the PyTorch development branch, but the next release is about 2 months off. 

## Reference to official issue

https://github.com/concept-to-clinic/concept-to-clinic/issues/4
https://github.com/concept-to-clinic/concept-to-clinic/issues/1

## Motivation and Context

This Pull request implements  part of the prediction service (https://concept-to-clinic.readthedocs.io/en/latest/design-doc.html#detect-prediction-service) 

## How Has This Been Tested?

The algorithm has largely been tested by manually confirming about 5 random LUNA2016 datasets - as the loaded model has already been tested extensively elsewhere, this was deemed sufficient. 
It has been tested both outside of docker, on a Titan X GPU, and inside docker on an i7 Intel CPU.

## CLA
- [X] I have signed the CLA; if other committers are in the commit history, they have signed the CLA as well

### CODE REVIEW

File: prediction/src/tests/test_identify_trained_model_predict.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+from ..algorithms.identify import trained_model
+
+@pytest.fixture
+def dicom_path():
+    yield '../images/LIDC-IDRI-0001/1.3.6.1.4.1.14519.5.2.1.6279.6001.298806137288633453246975630178/' \
+          '1.3.6.1.4.1.14519.5.2.1.6279.6001.179049373636438705059720603192'
+
+
+
+def test_classify_predict_model_load(dicom_path):
```
Comment (WGierke) [inline prediction/src/tests/test_identify_trained_model_predict.py]: As @lamby already stated [here](https://github.com/concept-to-clinic/concept-to-clinic/pull/118#issuecomment-330149196) it'd be nice to still have some tests that validate the output of the models. I'm currently working on some tags to only execute these long running tests if one explicitly desires to do so. So feel free to write this test a bit more detailed :)

**Comment by isms** [comment]
> [...] the pull request also disables the Keras import.

@dchansen I'm not sure this is a feasible compromise. Is there a way to reimplement this in Tensorflow?

File: prediction/src/tests/test_identify_trained_model_predict.py (hunk 1 above)
Comment (dchansen) [reply prediction/src/tests/test_identify_trained_model_predict.py]: Sure, I'll have a look at that tomorrow.

**Comment by dchansen** [comment]
@isms I agree that it's not a feasible compromise at all. I did it mainly to highlight the issue.

I think we should have a single deep learning framework for the project as a whole, as they don't really play well together. I am not sure if Keras is the best choice, but either way, the project leads should make the decisions. Perhaps we should open an issue?

Converting the model itself into Keras would relatively easy, but converting the weights is somewhat more involved. Converting to Caffe2 or CNTK would be easy, as they all support the ONNX format.

File: prediction/src/tests/test_identify_trained_model_predict.py (hunk 1 above)
Comment (dchansen) [reply prediction/src/tests/test_identify_trained_model_predict.py]: I updated the test cases to be a little more comprehensive.

**Comment by dchansen** [comment]
I am a little unsure why, but the conflict between Tensorflow and Pytorch seems gone. It was realted to this issue in Tensorflow https://github.com/tensorflow/models/issues/523

**Comment by lamby** [comment]
This should block on #118 due to the slow test stuff.

**Comment by reubano** [comment]
Looking pretty good! Just a few questions for you. I noticed that in `classify/trained_model/predict`, you commented out the previous code instead of deleted it. Was just curious, why did you decided to keep it around? Also there's still some spacing errors. Can you run `flake8 prediction` in your terminal? Finally, I see this travis error:

```bash
$ git lfs pull
Git LFS: (540 of 540 files, 2 skipped) 497.08 MB / 497.08 MB, 41.20 MB skipped 
[eb2c037dd55e2f49da95657f7a7851cbcfe2f2b516848ed03f8c5c820f3e16b4] Object does not exist on the server: [404] Object does not exist on the server
[3a46182c1c4883e170d5b576d3a2645309f34817c53c16d4f7f29fbdcf799308] Object does not exist on the server: [404] Object does not exist on the server
```

I'll try to pinpoint which files exactly are missing. Meanwhile, can you verify that the model you added to `git-lfs` was properly uploaded? Thanks for your work!

**Comment by dchansen** [comment]
The git lfs issue should be resolved now. 
I kept the previous code because I was unsure what the long-term plans were regarding the style of the project. Conceivably, this pull request could be changed to fit the style of the previous one. 
While I don't agree with all the choices made, it was in many ways nicer than what I contributed here.  The main reason I did not follow it more closely was one of time, as I just wanted a working adaption of the gtr123 model as soon as possible.

I have fixed all the flake8 issues, except the complexity ones. Hopefully, I will have them fixed by the end of the day.

**Comment by reubano** [comment]
> The git lfs issue should be resolved now.

Awesome!

> I kept the previous code because I was unsure what the long-term plans were regarding the style of the project. Conceivably, this pull request could be changed to fit the style of the previous one. While I don't agree with all the choices made, it was in many ways nicer than what I contributed here. The main reason I did not follow it more closely was one of time, as I just wanted a working adaption of the gtr123 model as soon as possible.

Ok, I understand. I'm taking a look to see how to integrate both versions.

> I have fixed all the flake8 issues, except the complexity ones. Hopefully, I will have them fixed by the end of the day.

Great!

**Comment by reubano** [comment]
Ok... here's my attempt. What do you think?

```python
import SimpleITK as sitk
from src.algorithms.classify.src import gtr123_model
from src.preprocess.load_ct import load_ct, MetaData


def predict(dicom_path, centroids, model_path=None,
            preprocess_ct=None, preprocess_model_input=None):
    reader = sitk.ImageSeriesReader()
    filenames = reader.GetGDCMSeriesFileNames(dicom_path)

    if not filenames:
        raise FileNotFoundError()

    reader.SetFileNames(reader.GetGDCMSeriesFileNames(dicom_path))
    image, meta = reader.Execute(), None

    if preprocess_ct:
        meta = meta or load_ct(dicom_path)[1]
        voxel_data = preprocess_ct(image, MetaData(meta))
    else:
        voxel_data = image

    if preprocess_model_input:
        preprocessed = preprocess_model_input(voxel_data, centroids)
    else:
        preprocessed = voxel_data

    return gtr123_model.predict(preprocessed, centroids)
```

Since the keras model is now gone, do you see any reason to continue to keeping `model_path`? If not, I can help you refactor it out.

**Comment by dchansen** [comment]
This would work for now.
I think the long term solution should be to move the gtr123 preprocessing into preprocess_ct. I will probably not have time to do that until next week though.

**Comment by reubano** [comment]
> I think the long term solution should be to move the gtr123 preprocessing into preprocess_ct.

Agreed. However, I think your PR can work without that at this stage (MVP). I looked around some more and saw that `gtr123_model.predict` actually takes a `model_path` param. So the last line could be `return gtr123_model.predict(preprocessed, centroids, model_path)`. I think then, the only change would have to be [#L14-L16 here](https://github.com/concept-to-clinic/concept-to-clinic/blob/44d50164488adb9c84f1b98824854c53ccc12181/prediction/src/tests/test_classify_trained_model_predict.py#L14-L16) in `test_classify_trained_model_predict.py`.

**Comment by reubano** [comment]
Looking good! What are your thoughts about incorporating something like I [mentioned here](https://github.com/concept-to-clinic/concept-to-clinic/pull/122#issuecomment-331403903)? Main reason is just so that we don't disable current functionality with this PR. I think with that, we can get this merged.

**Comment by dchansen** [comment]
I'm trying to do a reasonable merge right now, and algorith.identify.trained_model is giving me a headache.
I think we really need a standardized way of specifying a model, which is independent of frameworks (Keras, Torch, Tensorflow, etc).

I would suggest that each model independently implements a predict method taking only a dicom path (for identification) or a dicom path and a list of candidates. trained_model.predict would then take the same argument, plus a string identifying the algorithm to be used .  How does this sound?

**Comment by vessemer** [comment]
@dchansen I'm right now working on that, also I've tried to convert the grt123 classification model into Keras (TF), but I've faced troubles with the `BatchNorm` layer, despite it transfer weights correctly it feeds me with different results (if I just throw away each `BatchNorm` from the graph then the output becomes just identical, thus I need to dig deeper).
What is related to the way of standardization without transferring the model I've almost done.

**Comment by dchansen** [comment]
@vessemer I am a little unsure about the benefit of moving the model to Keras now that Tensorflow and Torch seem to work without disturbing one another. Different approaches will require different enough post- and pre-processing  that they can't be represented as a Keras model anyway. 
Long term, I do see the advantage of having all models on a single deep-learning framework, but is Keras the right choice? Using [ONNX](https://github.com/onnx/onnx) as the way to save the model would allow for easy use of many frameworks, including Caffe2 and CNTK. 
Either way, as soon as there is a consensus, I would be happy to help converting the models to that particular format.

One difference between the Torch and Keras batchnorm is that they use a different epsilon value (1e-5 vs 1e-3), so maybe that's the issue? 
Could you make your WIP available on the model standardization?

**Comment by dchansen** [comment]
@reubano Yes, that was my thought exactly. The problem right now is that all the models take ownership of trained_model.predict, and there is no good way of switching between particular models. Ideally, we might also like to do meta-model, combining many different ones, and a uniform interface would help greatly.

**Comment by vessemer** [comment]
@dchansen, thank you for your input, but I've taken into account the difference of the default values, I've used the [nn-transfer](https://github.com/gzuidhof/nn-transfer) for the model standardization.

**Comment by dchansen** [comment]
@vessemer Well, if you can share the code, I can have a look at it.

**Comment by reubano** [comment]
> The problem right now is that all the models take ownership of trained_model.predict, and there is no good way of switching between particular models. Ideally, we might also like to do meta-model, combining many different ones, and a uniform interface would help greatly.

For now, I think this is the preferred course of action. We want to focus on one main model for each of (classify, identify, and segment). Later on, we may consider making it easier to switch out the models. The current area of flexibility is in the serialized trained data, e.g., `assets/dsb2017_detector.ckpt`. So while it is still tied to the underlying python ML model, you can tweak the training parameters as needed. See [my comment on #2](https://github.com/concept-to-clinic/concept-to-clinic/issues/2#issuecomment-325328384) for more info.

Also, let me know what exactly is giving you a headache so I can try and help you out.

**Comment by dchansen** [comment]
My issue was that I had to disable a lot of code  in[identify.trained_model](https://github.com/dchansen/concept-to-clinic/blob/3396244f45f6683bdcc74b8f05081c16549ae315/prediction/src/algorithms/identify/trained_model.py#L22-L71)
But I think this is what you want?

**Comment by reubano** [comment]
I was literally suggesting you just copy/pasted [my code from here](https://github.com/concept-to-clinic/concept-to-clinic/pull/122#issuecomment-331403903).

File: prediction/src/algorithms/classify/trained_model.py [hunk 2]
```diff
@@ -43,23 +43,24 @@ def predict(dicom_path, centroids, model_path=None,
...
+    else:
+        voxel_data = image
 
-    return centroids
+    if preprocess_model_input:
+        preprocessed = preprocess_model_input(voxel_data, centroids)
+    else:
+        preprocessed = voxel_data
+
+    return gtr123_model.predict(preprocessed, centroids)
```
Comment (reubano) [inline prediction/src/algorithms/classify/trained_model.py]:
```python
...
model_path = model or 'src/algorithms/classify/assets/gtr123_model.ckpt'
...
return gtr123_model.predict(image, centroids, model_path)
```

**Comment by reubano** [comment]
Just about there! Does this solve your headache with `algorithms.identify.trained_model`?

File: prediction/src/algorithms/classify/trained_model.py (hunk 2 above)
Comment (dchansen) [reply prediction/src/algorithms/classify/trained_model.py]: Not sure what you're asking here?

Comment (reubano) [reply prediction/src/algorithms/classify/trained_model.py]: Just reintroducing the `model_path` part [from here](https://github.com/concept-to-clinic/concept-to-clinic/pull/122/commits/ddcb4528777c4c097de823dad86f8708397fa1fe)

**Comment by reubano** [comment]
From the travis log

```bash
$ flake8 prediction
prediction/src/algorithms/identify/trained_model.py:13:1: F401 'glob' imported but unused
prediction/src/algorithms/identify/trained_model.py:15:1: F401 'dicom' imported but unused
prediction/src/algorithms/identify/trained_model.py:16:1: F401 'src.preprocess.errors.EmptyDicomSeriesException' imported but unused
prediction/src/algorithms/identify/trained_model.py:17:1: F401 'src.preprocess.lung_segmentation.save_lung_segments' imported but unused
prediction/src/algorithms/classify/trained_model.py:68:56: E231 missing whitespace after ','
prediction/src/preprocess/extract_lungs.py:57:19: E127 continuation line over-indented for visual indent
prediction/src/preprocess/extract_lungs.py:66:18: E127 continuation line over-indented for visual indent
prediction/src/preprocess/extract_lungs.py:67:18: E127 continuation line over-indented for visual indent
prediction/src/preprocess/extract_lungs.py:68:18: E127 continuation line over-indented for visual indent
prediction/src/tests/test_endpoints.py:10:1: F811 redefinition of unused 'json' from line 7
The command "flake8 prediction" exited with 1.

$ pycodestyle prediction
prediction/src/algorithms/classify/trained_model.py:68:56: E231 missing whitespace after ','
prediction/src/preprocess/extract_lungs.py:57:19: E127 continuation line over-indented for visual indent
prediction/src/preprocess/extract_lungs.py:66:18: E127 continuation line over-indented for visual indent
prediction/src/preprocess/extract_lungs.py:67:18: E127 continuation line over-indented for visual indent
prediction/src/preprocess/extract_lungs.py:68:18: E127 continuation line over-indented for visual indent
```

**Comment by lamby** [comment]
@dchansen This now has conflicts with ``master`` — could you rebase? :)

File: prediction/src/tests/test_identify_trained_model_predict.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import os
+
+import numpy as np
+import pytest
+
+from ..algorithms.identify import trained_model
+
+
+def skip_slow_test():
```
Comment (WGierke) [inline prediction/src/tests/test_identify_trained_model_predict.py]: Sorry for giving feedback at such a late state but you're not supposed to re-implement this function as it was already added to [prediction/src/tests/test_endpoints.py](https://github.com/concept-to-clinic/concept-to-clinic/blob/757dfea9959db8e79f9b4ab5c04748aea711ca77/prediction/src/tests/test_endpoints.py) :) Are you able to import it from there?

**Comment by dchansen** [comment]
@lamby Github claims that it has no conflicts with the base branch?

**Comment by reubano** [comment]
Still some failing tests... travis doesn't show details however, so I'm looking into it locally.

File: prediction/src/tests/test_identify_trained_model_predict.py (hunk 3 above)
Comment (reubano) [reply prediction/src/tests/test_identify_trained_model_predict.py]: nice catch @WGierke!

**Comment by dchansen** [comment]
The test run in my local environment (outside docker). Inside docker they run when individually, but not when running them all at once. I suspect it is tensorflow/torch shenanigans, but it is quite tricky to debug.

**Comment by dchansen** [comment]
Using a different docker image solved the problem for me.
I cannot for the life of me figure out where the segfault is coming from in the python3:6 image however.

**Comment by reubano** [comment]
Yea that is odd... which image works? And what is the one that doesn't?

**Comment by dchansen** [comment]
`pytorch:36` segfaults. 
`varunagrawal/pytorch` does not, but I had to manually install opencv and tkinter in it.

**Comment by dchansen** [comment]
I changed the image to be Ubuntu based instead. All tests should pass now.

**Comment by reubano** [comment]
Ok.... looks like the tests passed, but there are some new lint errors now

```bash
$ flake8 prediction
prediction/src/tests/__init__.py:1:1: F401 'torch' imported but unused
prediction/src/tests/__init__.py:1:13: E261 at least two spaces before inline comment
prediction/src/tests/__init__.py:1:14: E262 inline comment should start with '# '
The command "flake8 prediction" exited with 1.

$ pycodestyle prediction
prediction/src/tests/__init__.py:1:13: E261 at least two spaces before inline comment
prediction/src/tests/__init__.py:1:14: E262 inline comment should start with '# '
The command "pycodestyle prediction" exited with 1.
```

**Comment by dchansen** [comment]
The torch import is required simply because torch needs to be imported before Keras, always. I am very open to other ways of doing that.

**Comment by reubano** [comment]
Try adding this on the offending line: `# noqa # pylint: disable=unused-import`

**Comment by reubano** [comment]
Looks like everything is a go now! Do you mind squashing to 1 commit?

**Comment by dchansen** [comment]
I seem to have messed that up rather thoroughly. This will take me a while to sort out.

**Comment by reubano** [comment]
No worries :). Perhaps I can help out. What seems to be the issue(s)?

**Comment by dchansen** [comment]
Just not very experienced in using git rebase. I found myself in the situation of having to redo every merge made on the project for the last ten days. 
I have messed up my repository quite badly - I think my best approach is to try again from home, where I have a pristine version of my repo, before my attempts at rebasing.

**Comment by WGierke** [comment]
Instead of rebasing, I prefer to squash the commits like [so](https://gist.github.com/patik/b8a9dc5cd356f9f6f980). It fulfills the purpose as well, namely having only 1 commit in the end :)

**Comment by reubano** [comment]
You can also take a look at the [advice I posted here](https://github.com/concept-to-clinic/concept-to-clinic/pull/55#issuecomment-325956598).

**Comment by reubano** [comment]
Also `git config --global rerere.enabled true` will save conflict resolutions so in that in the future, you only have to resolve them once.

**Comment by dchansen** [comment]
@WGierke Thank you! Uploading the squashed commit now.

**Comment by reubano** [comment]
```bash
$ flake8 interface
interface/backend/api/tests.py:44:1: W293 blank line contains whitespace
interface/backend/api/tests.py:48:57: E502 the backslash is redundant between brackets
interface/backend/api/tests.py:49:81: E502 the backslash is redundant between brackets
interface/backend/api/serializers.py:74:1: E302 expected 2 blank lines, found 1
interface/backend/api/serializers.py:120:17: E127 continuation line over-indented for visual indent
interface/backend/api/views.py:24:1: F811 redefinition of unused 'Response' from line 13
interface/backend/api/views.py:47:37: W291 trailing whitespace
The command "flake8 interface" exited with 1.

$ pycodestyle interface
interface/backend/api/serializers.py:74:1: E302 expected 2 blank lines, found 1
interface/backend/api/serializers.py:120:17: E127 continuation line over-indented for visual indent
interface/backend/api/tests.py:44:1: W293 blank line contains whitespace
interface/backend/api/tests.py:48:57: E502 the backslash is redundant between brackets
interface/backend/api/tests.py:49:81: E502 the backslash is redundant between brackets
interface/backend/api/views.py:47:37: W291 trailing whitespace
The command "pycodestyle interface" exited with 1.
```

**Comment by dchansen** [comment]
But the pull request doesn't include those files??

**Comment by isms** [comment]
@reubano @dchansen Those issues were introduced by recent commits unrelated to this PR. I pushed fixes which should get the code style back to passing. Will restart the build on this branch to update the status.

(cc @lamby so we can avoid in the future!)

**Comment by reubano** [comment]
Ok... master looks good. You mind rebasing @dchansen to ensure everything is good to go?

**Comment by reubano** [comment]
closes #1 and #4 :shipit:

**Pull request merged**